controllers/volumereplicationgroup_controller.go

```diff
@@ -1235,19 +1234,59 @@ func (v *VRGInstance) undoPVRetentionForPVC(pvc corev1.PersistentVolumeClaim, lo
 
 // Upload PV to the list of S3 stores in the VRG spec
 func (v *VRGInstance) uploadPVToS3Stores(pvc *corev1.PersistentVolumeClaim, log logr.Logger) (err error) {
+	// Find the ProtectedPVC of the given PVC in v.instance.Status.ProtectedPVCs[]
+	protectedPVC := v.findProtectedPVC(pvc.Name)
+	// Find the ClusterDataProtected condition of the given PVC in ProtectedPVC.Conditions
+	clusterDataProtected := findCondition(protectedPVC.Conditions, VRGConditionTypeClusterDataProtected)
+
+	// Optimization: skip uploading the PV of this PVC if it was uploaded previously
+	if clusterDataProtected != nil && clusterDataProtected.Status == metav1.ConditionTrue &&
+		clusterDataProtected.ObservedGeneration == v.instance.Generation {
+		// v.log.Info("PV cluster data already protected")
+		return nil
+	}
+
+	s3Profiles := []string{}
+	// Upload the PV to all the S3 profiles in the VRG spec
 	for _, s3ProfileName := range v.instance.Spec.S3ProfileList {
 		if err := v.reconciler.PVUploader.UploadPV(v, s3ProfileName, pvc); err != nil {
-			msg := "Failed to upload PV cluster data to s3Profile " + s3ProfileName
-			v.updateProtectedPVCDataCondition(pvc.Name, VRGConditionReasonError, msg)
+			msg := fmt.Sprintf("Error uploading PV cluster data to s3Profile %s, %v",
+				s3ProfileName, err)
+			v.updatePVCClusterDataProtectedCondition(pvc.Name, VRGConditionReasonUploadFailed, msg)
 			log.Error(err, msg)
 			rmnutil.ReportIfNotPresent(v.reconciler.eventRecorder, v.instance, corev1.EventTypeWarning,
 				rmnutil.EventReasonPVUploadFailed, err.Error())
 
-			return fmt.Errorf("error uploading PV %s using profile %s, err %w", pvc.Name, s3ProfileName, err)
+			return fmt.Errorf("error uploading cluster data of PV %s to S3 profile %s, %w",
+				pvc.Name, s3ProfileName, err)
 		}
+
+		// Successfully uploaded to S3ProfileName
+		s3Profiles = append(s3Profiles, s3ProfileName)
```

Comment:
(nit) We error out if we fail to upload to any s3Profile. So s3Profiles at the end of the loop should contain all profiles in `v.instance.Spec.S3ProfileList`. Based on this we can error out up front the case of no s3 profiles found (say at line 1249), and then at 1269 we can drop the if.

As stated just a nit, if we merge this before we can leave it as is.

Reply:
Good point.  I will fix it.  Given that we error out if there are no S3 profiles, I believe that the code that tolerates a VRG without any S3 profiles can be removed. The intent behind such tolerance was to allow creating a VRG that for the sake of creating VRs without the need to upload PVs.  If such a mode is required, it is perhaps best to allow such tolerance with a spec field in future.

Reply:
I pushed a new commit for ease of review but can be optionally squashed when merging.